server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java

```diff
@@ -392,4 +397,21 @@ public static boolean descendsFromGlobalAggregator(Aggregator parent) {
         return false;
     }
 
+    @Override
+    protected void preGetSubLeafCollectors(LeafReaderContext ctx) throws IOException {
```

Comment:
I think we should move this to `doPreCollection()` if possible.  That will get called once before starting to collect docs on the entire shard, whereas `preGetSubLeafCollectors()` is called once per segment in the shard

If it's not possible to move for some reason, we should add a `if (fieldDocCountProvider != null) { return } ` check so we only load it once

Reply:
`FieldBasedDocCountProvider` requires the instance of the `LeafReaderContext` for reading the `doc_count` field. 

In commit c0f23ae I separated the code that resolves the `DocCountProvider` (and put this part in the constructor) from loading the `LeafReaderContext` and put it in the `preGetSubLeafCollectors()` method.

Reply:
Looks great, I like this modification a lot! :+1:   I think the extra method on the interface is worth it in this scenario